sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala

```diff
@@ -641,12 +641,20 @@ class GpuOrcPartitionReader(
       }
       isFirstBatch = false
     }
-    batch.isDefined
+
+    batch.map(_ => true).getOrElse {
+      // After finishing reading, we should clean up ctx early, just in case leaking orc readers
+      cleanUpOrc(ctx)
+      isOrcCleanedUp = true
+      false
+    }
   }
 
   override def close(): Unit = {
     super.close()
-    cleanUpOrc(ctx)
+    if (!isOrcCleanedUp) {
+      cleanUpOrc(ctx)
```

Comment:
I'm assuming this check is here because cleanupOrc is not idempotent.  Should this flag be part of the context and checked within `cleanUpOrc` rather than requiring callers to track this context-specific state separately?

Reply:
Done